zkevm-circuits/src/evm_circuit/execution/error_oog_codestore.rs

```diff
@@ -0,0 +1,221 @@
+    opcode: Cell<F>,
+    is_create: Cell<F>,
+    memory_address: MemoryAddressGadget<F>,
+    code_store_gas_insufficient: LtGadget<F, N_BYTES_GAS>,
+    restore_context: RestoreContextGadget<F>,
+}
+
+impl<F: Field> ExecutionGadget<F> for ErrorOOGCodeStoreGadget<F> {
+    const NAME: &'static str = "ErrorOutOfGasCodeStore";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::ErrorOutOfGasCodeStore;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+        cb.opcode_lookup(opcode.expr(), 1.expr());
+
+        let offset = cb.query_cell_phase2();
+        let length = cb.query_word_rlc();
+        cb.stack_pop(offset.expr());
+        cb.stack_pop(length.expr());
+        let memory_address = MemoryAddressGadget::construct(cb, offset, length);
+
+        let is_create = cb.call_context(None, CallContextFieldTag::IsCreate);
+        cb.require_true("is_create is true", is_create.expr());
+
+        // constrain code store gas > gas left, that is GasCost::CODE_DEPOSIT_BYTE_COST
+        // * length > gas left
+        let code_store_gas_insufficient = LtGadget::construct(
+            cb,
+            cb.curr.state.gas_left.expr(),
+            GasCost::CODE_DEPOSIT_BYTE_COST.expr() * memory_address.length(),
+        );
+        cb.require_equal(
+            "CODE_DEPOSIT_BYTE_COST * length > step gas left",
+            code_store_gas_insufficient.expr(),
+            1.expr(),
+        );
+
+        // restore context as in internal call
+        cb.require_zero("in internal call", cb.curr.state.is_root.expr());
+
+        // Case C in the return specs.
+        let restore_context = RestoreContextGadget::construct(
+            cb,
+            0.expr(),
+            0.expr(),
+            memory_address.offset(),
+            memory_address.length(),
+            0.expr(),
+            0.expr(),
+        );
+
+        Self {
+            opcode,
+            is_create,
+            memory_address,
+            code_store_gas_insufficient,
+            restore_context,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _tx: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        self.opcode
+            .assign(region, offset, Value::known(F::from(opcode.as_u64())))?;
+
+        let [memory_offset, length] = [0, 1].map(|i| block.rws[step.rw_indices[i]].stack_value());
+        self.memory_address
+            .assign(region, offset, memory_offset, length)?;
+
+        self.is_create
+            .assign(region, offset, Value::known(F::from(call.is_create as u64)))?;
+        self.code_store_gas_insufficient.assign(
+            region,
+            offset,
+            F::from(step.gas_left),
+            F::from(200 * length.as_u64()),
```

Comment:
make 200 a constant?

Reply:
updated in [fbea334](https://github.com/scroll-tech/zkevm-circuits/pull/304/commits/fbea3347b5e799723a89c9a3c8e2b5eb05579c22) thanks!